apps/votes: added widget to voting form

sabinammm opened the pull request:
This produces three instead of one input field for the voting token. Each field requires 4 characters.

Still to do

- [x] Form styling 
- [x] The form validation is in contradiction and requires changes, see https://github.com/liqd/a4-meinberlin/pull/4033#discussion_r763040554

@phillimorland would you check out the styling when you get the chance? 
@Rineee I added comments for the form validation that I can't get to work well, maybe you can check if you have time? I think the decompress and value_from_dict functions work.

sabinammm commented on meinberlin/apps/votes/forms.py:
```diff
@@ -4,9 +4,29 @@
+
+
+class TokenForm(forms.Form):
+    token = forms.CharField(max_length=40, widget=VotingTokenWidget())
```
The max_length validation doesn't work well with the three fields if it is shortened to 4. When it is shortened, each field really only allows 4 characters to be entered, but since we have a total of 12, this is in contradtion.

sabinammm commented on meinberlin/apps/votes/forms.py:
```diff
@@ -4,9 +4,29 @@
+
+    def __init__(self, attrs=None):
+        widgets = (
+            forms.TextInput(attrs={'size': '4', 'maxlength': '4'}),
```
While this limits the characters actually used by the field, the styling still looks aweful and this contradcits the max_length=40 defined in line 29

Rineee commented:
@sabinammm You were right that the maxlength from the widgets were ignored, because of the maxlength on the form field.. I found [this](https://stackoverflow.com/questions/49349907/django-doesnt-respect-the-maxlength-widget-attribute) issue, that wasnt answered, but pointed me in the right direction. I think it shouldnt be like that, but anyway it is and I now removed the maxlength from the form field to make it work. I also added minlength to the widgets, not sure if we then also need size or is that for styling, do you know?
I added the changes in another commit, so you could look at them, maybe you could squash before merging?

Rineee commented on meinberlin/apps/votes/forms.py:
```diff
@@ -4,9 +4,35 @@
+
+    def decompress(self, value):
+        if value:
+            n = 4
```
I think you could remove this and simply write 4 in the line below?

sabinammm replied:
Yes, I thought about this, and then I wondered if we would ever change it and then it would be more explicit. But for it to be more explicit, it needs to be more of a telling variablename than simply 'n' I guess. I will change it! Thanks:)

sabinammm commented:
> @sabinammm You were right that the maxlength from the widgets were ignored, because of the maxlength on the form field.. I found [this](https://stackoverflow.com/questions/49349907/django-doesnt-respect-the-maxlength-widget-attribute) issue, that wasnt answered, but pointed me in the right direction. I think it shouldnt be like that, but anyway it is and I now removed the maxlength from the form field to make it work. I also added minlength to the widgets, not sure if we then also need the size attribute or is that for styling, do you know? I added the changes in another commit, so you could look at them, maybe you could squash before merging?

Yes, perfect!! Thanks!

phillimorland commented:
@sabinammm once you're happy with your code, let me know and i'll take over the styling :)

sabinammm commented:
> @sabinammm once you're happy with your code, let me know and i'll take over the styling :)

This is fine now and ready for styling. Thank you :)

phillimorland commented on meinberlin/apps/budgeting/templates/meinberlin_budgeting/proposal_list.html:
```diff
@@ -26,11 +26,11 @@
                 <div class="form-group">
                     {% csrf_token %}
-                    <div class="widget widget--{{ form.token|widget_type }}">
+                    <div class="widget widget--{{ form.token|widget_type }} widget--vote">
```
@sabinammm ok styling was more annoying then complicated but I had to add a class name here as there is nothing rendered from `widget_type` don't know why though?

phillimorland reviewed (approved):
very nice!
